test/net/sourceforge/kolmafia/persistence/ItemFinderTest.java

```diff
@@ -0,0 +1,833 @@
+  }
+
+  // ItemFinder is the utility which accepts a parameter string (as typed by a
+  // user in the gCLI) and parses it into an array of AdventureResults,
+  //
+  // The string can specify a single item or a comma seperated list of items.
+  //
+  // The user can specify a "count" for each item, including wildcards:
+  // "*" means "all available and "-N" means "all except N"
+
+  // Primitives:
+  //
+  // List<String> getMatchingNames(String)
+  // String getFirstMatchingItemName(List, String, Match)
+  // AdventureResult getFirstMatchingItem(String, boolean, List, Match)
+  // AdventureResult[] getMatchingItemList(String, boolean, List, Match)
+
+  // getMatchingItemList uses getFirstMatchingItem
+  // getFirstMatchingItem uses getFirstMatchingItemName
+  // getFirstMatchingItemName uses getMatchingNames
+  // getMatchingNames simply calls ItemDatabase.getMatchingNames
+
+  // The first thing to test is a string containing a single item
+  //
+  // "" -> ERROR
+  // "bogus" -> ERROR
+  // seal tooth
+  // 1 Meat
+  // mmj -> magical mystery juice
+  // 1337 7r0uZ0RZ
+  // Tea, Earl Grey, Hot
+  // [7961]Staff of Ed
+  // [7961]
+  //  ¶7961
+  //
+  // The "pilcrow" character followed by itemID was invented by Jason
+  // Harper to allow ASH RuntimeLibrary methods to call CLI commands
+  // ("buy", "closet", etc.) with no ambiguity.
+  //
+  // The [itemId] forms were added later as general disambiguation for
+  // multiple object types in ASH
+  //
+  // Item names can contain a comma. Even though the parameter string can
+  // contain multiple comma-separated items, we do not allow or expect you to
+  // escape the comma. Instead, you can only specify such an item (by name) if
+  // it is the entire parameter string.
+  //
+  // Although the user can specify N of an item, item names can start with a
+  // number. Tricky!
+  //
+  // Lastly notice that we do "fuzzy matching" using ItemDatabase.getMatchingNames(),
+  // which uses StringUtilities.getMatchingNames().
+  // That package deserves its own test suite.
+  //
+  // also:
+  //
+  // 3215 Meat -> MEAT, not an item
+
+  // *********************** getFirstMatchingItem ***********************
+  //
+  // That method assumes its parameter is the name of a single item,
+  // possible with a count. Assuming it unambiguously names a single
+  // item, it returns an AdventureResult with itemId and count.
+
+  @Test
+  public void itShouldParseSingleItem() {
+
+    // This set of tests will use null as the source list. That means
+    // that it will not restrict its matches to what is available in a
+    // specific list. The only exception is if you use wildcard counts -
+    // * or -N - in which case, it assumes inventory. We'll test wildcard
+    // counts separately.
+
+    AdventureResult item;
+    String message;
+    String parameter;
+
+    // Test successful exact match for entire string.
```

Comment:
If these were their own tests, you could have a name instead of a comment. You could use `@Nested class SingleItemParsing` to group all of them.

Not particularly fussed about this, I've never really agreed with the "each test should only test one thing" idea.

Reply:
> Not particularly fussed about this, I've never really agreed with the "each test should only test one thing" idea.

I've definitely found it help later down the line when you make a change and suddenly see a random test break - you know exactly what behaviour has changed.

Reply:
yeah, but if you look at the stack trace (or look at the assertion error message, if you provided one), you also know exactly which behaviour changed ;). You don't get to see multiple failures up-front, but given that you have to fix all the failures always anyway, it all comes out in the wash for me.

Reply:
Seeing multiple failures would have been helpful, rather than see one, fix it, run again, see next, etc.
Not familiar with @Nested.

As I mentioned in a previous comment, I want to rested the items lists before each group of tests (each test, as currently coded). And perhaps add items to the lists, custom per group of tests (or at the top of the function, as currently coded).

I want to get this in, so will study @Nested later - and try it out for my NEXT major test suite.
Or maybe I'll follow up in a different PR by refactoring this one, as a learning experience. :)

Reply:
Not changing structure at this time, as mentioned.